File: NeetCode/Two_Pointers/Merge_Sorted_Arrays/merge__sorted__arrays.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        m -= 1
        n -= 1
        for ptr1 in range(len(nums1) - 1, -1, -1):
            if m >= 0 and n >= 0:
                if nums1[m] >= nums2[n]:
                    nums1[ptr1] = nums1[m]
                    m -= 1
                else:
                    nums1[ptr1] = nums2[n]
                    n -= 1
            else:
                if n >= 0:
                    nums1[ptr1] = nums2[n]
                    n -= 1


if __name__ == "__main__":
    sol = Solution()
    nums1 = [10, 20, 20, 40, 0, 0]
    m = 4
    nums2 = [1, 2]
    n = 2
    sol.merge(nums1, m, nums2, n)
    print(nums1)
    print("Running Solution...")

File: NeetCode/Two_Pointers/Merge_Sorted_Arrays/Merge_Sorted_Arrays.cpp
// Merge_Sorted_Arrays.cpp
//
// Compile locally with:
// clang++ Merge_Sorted_Arrays.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    void merge(vector<int> &nums1, int m, vector<int> &nums2, int n)
    {
        m--;
        n--;
        for (int index = nums1.size(); index >= 0; index--)
        {
            if (m >= 0 && n >= 0)
            {
                if (nums1[m] >= nums2[n])
                {
                    nums1[index] = nums1[m];
                    m--;
                }
                else
                {
                    nums1[index] = nums2[n];
                    n--;
                }
            }
            else
            {
                if (n >= 0)
                {
                    nums1[index] = nums2[n];
                    n--;
                }
            }
        }
    }
};

int main()
{
    cout << "Running Merge_Sorted_Arrays.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
